### PR TITLE
ZAM/ZAT/ZAS restructure

### PR DESCRIPTION
**Pull request by ocke**
:koala:

/cc @zendesk/quokka

### References
 - Jira link: 

### Risks
- Refactoring always entails risks

### CODE REVIEW

File: lib/zendesk_apps_support/assets/installed.js.erb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+(function() {
```
Comment (danielbreves) [inline lib/zendesk_apps_support/assets/installed.js.erb]: Can we use ZAS to generate `installed.js` for ZAM too?

File: lib/zendesk_apps_support/assets/installed.js.erb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  <% appsjs.each do | appjs | %>
+    <%= appjs %>
+  <% end %>
+
+  <% installationsjs.each do | installationjs | %>
+    <%= installationjs %>
```
Comment (danielbreves) [inline lib/zendesk_apps_support/assets/installed.js.erb]: Any reason why this needs to be in a separate template?

File: lib/zendesk_apps_support/assets/installed.js.erb (hunk 1 above)
Comment (ocke) [inline lib/zendesk_apps_support/assets/installed.js.erb]: done.

File: lib/zendesk_apps_support/assets/installed.js.erb (hunk 2 above)
Comment (ocke) [inline lib/zendesk_apps_support/assets/installed.js.erb]: nope, there was first, with how I structured it, but not anymore I guess.

File: lib/zendesk_apps_support/assets/installed.js.erb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+(function() {
+  <% appsjs.each do | appjs | %>
```
Comment (adammw) [inline lib/zendesk_apps_support/assets/installed.js.erb]: nit: spaces between the pipes looks very foreign to me.

Comment (itirkarp) [inline lib/zendesk_apps_support/assets/installed.js.erb]: lol +1 it took a second to figure out what this meant. But that could also be the vacation

File: lib/zendesk_apps_support/assets/src.js.erb [hunk 4]
```diff
@@ -1,6 +1,5 @@
-(function() {
-  with( ZendeskApps.AppScope.create() ) {
-    <% unless modules.nil? %>
+with( ZendeskApps.AppScope.create() ) {
+  <% unless modules.empty? %>
```
Comment (adammw) [inline lib/zendesk_apps_support/assets/src.js.erb]: `if modules.present?` ?

File: lib/zendesk_apps_support/package.rb [hunk 5]
```diff
@@ -46,12 +60,24 @@ def validate
...
     end
 
+    # This -production- function is used after unzipping the zip file in ZAM.
+    # Shouldn't be used for -dev-
+    def validate!
+      if Dir["#{@root}/**/{*,.*}"].any? { |f| File.symlink?(f) }
+        raise ValidationError.new('Symlinks are not allowed in the zip file')
+      end
+
+      errors = self.validate
```
Comment (adammw) [inline lib/zendesk_apps_support/package.rb]: self?

File: lib/zendesk_apps_support/installed.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+require 'erubis'
+
+module ZendeskAppsSupport
+  class Installed
+    INSTALLED_TEMPLATE = Erubis::Eruby.new( File.read(File.expand_path('../assets/installed.js.erb', __FILE__)) )
```
Comment (adammw) [inline lib/zendesk_apps_support/installed.rb]: Do we need to use File.join or anything for windows compatibility?

File: lib/zendesk_apps_support/package.rb [hunk 7]
```diff
@@ -93,20 +202,64 @@ def requirements_json
...
+      remove_zendesk_keys(translations)
     end
 
     def app_translations(locale)
-      remove_zendesk_keys(translations(locale))
+      remove_zendesk_keys(translations[locale])
     end
 
-    def readified_js(app_name, app_id, asset_url_prefix, settings = {}, locale = 'en')
+    def readified_js(app_id, asset_url_prefix, locale = 'en')
```
Comment (adammw) [inline lib/zendesk_apps_support/package.rb]: I wonder if we should use a hash here rather than relying on us knowing the argument order.

File: lib/zendesk_apps_support/package.rb [hunk 8]
```diff
@@ -93,20 +202,64 @@ def requirements_json
...
       manifest = manifest_json
       source = app_js
-      name = app_name || manifest[:name] || 'Local App'
```
Comment (adammw) [inline lib/zendesk_apps_support/package.rb]: don't we need this?

File: lib/zendesk_apps_support/package.rb [hunk 9]
```diff
@@ -131,11 +282,9 @@ def readified_js(app_name, app_id, asset_url_prefix, settings = {}, locale = 'en
           asset_url_prefix: asset_url_prefix,
           app_class_name: app_class_name,
           author: author,
-          translations: app_translations(locale),
+          translations: translations[locale],
```
Comment (adammw) [inline lib/zendesk_apps_support/package.rb]: is there much of a difference between app_translations and translations? the app_translations method still exists.

File: lib/zendesk_apps_support/package.rb [hunk 10]
```diff
@@ -173,8 +322,6 @@ def file_path(path)
       File.join(root, path)
     end
 
-    private
-
     def compiled_templates(app_id, asset_url_prefix)
       compiled_css = ZendeskAppsSupport::StylesheetCompiler.new(DEFAULT_SCSS + customer_css, app_id, asset_url_prefix).compile
```
Comment (adammw) [inline lib/zendesk_apps_support/package.rb]: the block within (the `templates = ...`) is already a method (`templates`)

**Comment by adammw** [comment]
Are we moving some of the deleted specs/tests from ZAM here?

**Comment by adammw** [comment]
This is awesome, I've been trying to get this de-duplicated for ages :clap:

File: lib/zendesk_apps_support/assets/installed.js.erb (hunk 3 above)
Comment (adammw) [inline lib/zendesk_apps_support/assets/installed.js.erb]: i would normally link to a rubocop rule but I couldn't find one

File: lib/zendesk_apps_support/installed.rb (hunk 6 above)
Comment (ocke) [inline lib/zendesk_apps_support/installed.rb]: We're already doing the same thing in package.rb.

File: lib/zendesk_apps_support/package.rb (hunk 8 above)
Comment (ocke) [inline lib/zendesk_apps_support/package.rb]: It moved to 110.

File: lib/zendesk_apps_support/package.rb (hunk 9 above)
Comment (ocke) [inline lib/zendesk_apps_support/package.rb]: yeah it filters out some things.

File: lib/zendesk_apps_support/assets/installed.js.erb [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+  <% installations.each do |installation| %>
+    ZendeskApps[<%= installation.app.name.to_json %>].install(<%= installation.to_json %>);
+  <% end %>
+
+  <% installation_orders.each do |loc_name, installation_list| %>
+    ZendeskApps.sortAppsForSite(<%= loc_name.to_json %>, <%= installation_list.to_json %>);
+  <% end %>
+}());
+
+ZendeskApps.trigger && ZendeskApps.trigger('ready');
```
Comment (svizzari) [inline lib/zendesk_apps_support/assets/installed.js.erb]: Not sure we need the guard here anymore, `ZendeskApps` will always be an event emitter now.

`ZendeskApps.trigger('ready')` should suffice.

Comment (danielbreves) [inline lib/zendesk_apps_support/assets/installed.js.erb]: Even in Zopim? :frog:

Comment (svizzari) [inline lib/zendesk_apps_support/assets/installed.js.erb]: Why won't/can't it be an event emitter in Zopim?

Comment (danielbreves) [inline lib/zendesk_apps_support/assets/installed.js.erb]: I think if we expect it to be an event emitter we should make it an event emitter in ZAF, rather than expecting the host to make it one. In Zopim we were actually using a callback queue: https://github.com/zopim/meshim-frontend/pull/1323/files#diff-f6efbb95e2a272f2ae751763b5547728R77

Comment (svizzari) [inline lib/zendesk_apps_support/assets/installed.js.erb]: I think we should make ZAF an event emitter.  We already have MicroEE there now.  Datanode can tell ZAF when to emit events.

File: lib/zendesk_apps_support/assets/src.js.erb (hunk 4 above)
Comment (ocke) [inline lib/zendesk_apps_support/assets/src.js.erb]: That's rails.

**Comment by adammw** [comment]
:+1: when :green_heart: / conflicts resolved

File: lib/zendesk_apps_support/validations/manifest.rb [hunk 12]
```diff
@@ -4,7 +4,7 @@ module ZendeskAppsSupport
   module Validations
     module Manifest
       REQUIRED_MANIFEST_FIELDS = %w( author defaultLocale ).freeze
-      OAUTH_REQUIRED_FIELDS    = %w( client_id client_secret authorize_uri access_token_uri ).freeze
+      OAUTH_REQUIRED_FIELDS    = %w( clientId clientSecret authorizeUri accessTokenUri ).freeze
```
Comment (danielbreves) [inline lib/zendesk_apps_support/validations/manifest.rb]: Can you split this out into a separate PR?

File: lib/zendesk_apps_support/package.rb [hunk 13]
```diff
@@ -65,58 +88,27 @@ def files
...
-      read_json(file_path, false)
-    end
-
-    def app_translations(locale)
-      remove_zendesk_keys(translations(locale))
-    end
-
-    def readified_js(app_name, app_id, asset_url_prefix, settings = {}, locale = 'en')
-      manifest = manifest_json
+    def compile_js(app_id, asset_url_prefix, locale = 'en')
```
Comment (danielbreves) [inline lib/zendesk_apps_support/package.rb]: Can you make these named params please?

File: config/locales/en.yml [hunk 14]
```diff
@@ -17,6 +17,7 @@ en:
               app.js
             manifest_not_json: manifest is not proper JSON. %{errors}
             missing_manifest: Could not find manifest.json
+            symlink_in_zip: Symlinks are not allowed in the zip file
```
Comment (sandlerr) [inline config/locales/en.yml]: https://github.com/zendesk/zendesk_apps_support/blob/master/config/locales/en.yml#L1 edit the other yml please

File: lib/zendesk_apps_support/package.rb [hunk 15]
```diff
@@ -42,22 +51,36 @@ def validate
...
-        errors.flatten!
+        errors.flatten!.compact!
       end
     end
 
-    def app_js
-      read_file('app.js')
+    def validate!(marketplace: true)
+      errors = validate(marketplace: marketplace)
+      raise errors.first if errors.any?
```
Comment (sandlerr) [inline lib/zendesk_apps_support/package.rb]: were we already only raising the first error? Would probably be good to somehow show all the information at once if possible.

File: lib/zendesk_apps_support/package.rb [hunk 16]
```diff
@@ -42,22 +51,36 @@ def validate
...
+      true
     end
 
-    def commonjs_modules
-      return unless has_lib_js?
+    def assets
+      @assets ||= begin
+                    pwd = Dir.pwd
+                    Dir.chdir(@root)
```
Comment (sandlerr) [inline lib/zendesk_apps_support/package.rb]: you can use 
```ruby
Dir.chdir(@root) do
  assets = Dir["assets/**/*"].select { |f| File.file?(f) }
end
```

for this entire block

File: lib/zendesk_apps_support/package.rb [hunk 17]
```diff
@@ -131,20 +132,131 @@ def readified_js(app_name, app_id, asset_url_prefix, settings = {}, locale = 'en
...
+    def requirements_json
+      return nil unless has_requirements?
+      @requirements ||= read_json('requirements.json')
+    end
+
+    def no_template
+      if manifest_json['noTemplate'].is_a?(Array)
+        false
+      else
+        !!manifest_json['noTemplate']
```
Comment (sandlerr) [inline lib/zendesk_apps_support/package.rb]: why do we need `!!` here?

File: lib/zendesk_apps_support/package.rb [hunk 18]
```diff
@@ -131,20 +132,131 @@ def readified_js(app_name, app_id, asset_url_prefix, settings = {}, locale = 'en
...
+      end
+    end
+
+    def compiled_templates(app_id, asset_url_prefix)
+      compiled_css = ZendeskAppsSupport::StylesheetCompiler.new(DEFAULT_SCSS + customer_css, app_id, asset_url_prefix).compile
+
+      templates = begin
+        Dir["#{root}/templates/*.hdbs"].inject({}) do |h, file|
```
Comment (sandlerr) [inline lib/zendesk_apps_support/package.rb]: can we give `h` a better name? `templates_hash`?

File: lib/zendesk_apps_support/package.rb [hunk 19]
```diff
@@ -131,20 +132,131 @@ def readified_js(app_name, app_id, asset_url_prefix, settings = {}, locale = 'en
...
+
+      layout = templates['layout'] || DEFAULT_LAYOUT.result
+
+      templates.tap do |templates|
+        templates['layout'] = "<style>\n#{compiled_css}</style>\n#{layout}"
+      end
+    end
+
+    def name
+      manifest_json['name'] || 'Local App'
```
Comment (sandlerr) [inline lib/zendesk_apps_support/package.rb]: people might be confused where this came from

File: lib/zendesk_apps_support/package.rb [hunk 20]
```diff
@@ -131,20 +132,131 @@ def readified_js(app_name, app_id, asset_url_prefix, settings = {}, locale = 'en
...
+    end
+
+    def market_translations(locale)
+      result = translations[locale].try(:[], 'app') || {}
+      result.delete('name')
```
Comment (sandlerr) [inline lib/zendesk_apps_support/package.rb]: alternately
`result.delete_if {|k| %w[name description long_description installation_instructions].include? k}`

Comment (sandlerr) [inline lib/zendesk_apps_support/package.rb]: btw, this does actually delete, did you want to dup?

File: lib/zendesk_apps_support/package.rb [hunk 21]
```diff
@@ -131,20 +132,131 @@ def readified_js(app_name, app_id, asset_url_prefix, settings = {}, locale = 'en
...
     def customer_css
       css_file = file_path('app.css')
       File.exist?(css_file) ? File.read(css_file) : ''
     end
 
+    private
+
+    def css
+      @css ||= begin
+                 File.read(@css_path)
```
Comment (sandlerr) [inline lib/zendesk_apps_support/package.rb]: this whitespace style looks odd, can we just indent once for the begin?

File: lib/zendesk_apps_support/package.rb [hunk 22]
```diff
@@ -131,20 +132,131 @@ def readified_js(app_name, app_id, asset_url_prefix, settings = {}, locale = 'en
...
+               rescue Errno::ENOENT
+                 ''
+               end
+    end
+
+    def templates
+      @templates ||= begin
+                       templates_dir = File.join(@root, 'templates')
```
Comment (sandlerr) [inline lib/zendesk_apps_support/package.rb]: same whitespace as above

File: lib/zendesk_apps_support/package.rb [hunk 23]
```diff
@@ -131,20 +132,131 @@ def readified_js(app_name, app_id, asset_url_prefix, settings = {}, locale = 'en
...
+    end
+
+    def translations
+      @translations ||= begin
+        translation_dir = File.join(@root, 'translations')
+        return {} unless File.directory?(translation_dir)
+
+        locale_path = "#{translation_dir}/#{self.manifest_json['defaultLocale']}.json"
+        default_translations = process_translations(locale_path)
+        Dir["#{translation_dir}/*.json"].inject({}) do |h, tr|
```
Comment (sandlerr) [inline lib/zendesk_apps_support/package.rb]: `h` and `tr` are pretty meaningless to me

File: lib/zendesk_apps_support/package.rb [hunk 24]
```diff
@@ -205,18 +296,40 @@ def non_tmp_files
...
+      end
+      result_h
+    end
+
     def read_file(path)
       File.read(file_path(path))
     end
 
-    def read_json(path, symbolize_names = true)
+    def read_json(path)
```
Comment (sandlerr) [inline lib/zendesk_apps_support/package.rb]: so why strings over symbols?

File: spec/package_spec.rb [hunk 25]
```diff
@@ -55,25 +55,24 @@
...
-      js = @package.readified_js(nil, 0, 'http://localhost:4567/')
+      js = @package.compile_js(app_id: 0, assets_dir: 'http://localhost:4567/0/')
 
       expected = <<HERE
-(function() {
-  with( ZendeskApps.AppScope.create() ) {
+with( ZendeskApps.AppScope.create() ) {
     require.modules = {
-        'a.js': function(exports, require, module) {
+        "a.js": function(exports, require, module) {
```
Comment (sandlerr) [inline spec/package_spec.rb]: hmm, what if an asset name has quotes in it?

Comment (ocke) [inline spec/package_spec.rb]: it get's put there by does .to_json, so it will escape double quote.
`"my\"file"`

File: lib/zendesk_apps_support/package.rb (hunk 24 above)
Comment (ocke) [inline lib/zendesk_apps_support/package.rb]: it was symbols in ZAT and strings in ZAM. I choose strings. Also Tobi told me it could be a security hole?

Comment (sandlerr) [inline lib/zendesk_apps_support/package.rb]: using symbols is a security hole?

Comment (sandlerr) [inline lib/zendesk_apps_support/package.rb]: if this is referring to lack of symbol GC that was fixed in MRI 2.2.

File: lib/zendesk_apps_support/package.rb [hunk 26]
```diff
@@ -173,25 +279,8 @@ def file_path(path)
       File.join(root, path)
```
Comment (danielbreves) [inline lib/zendesk_apps_support/package.rb]: Looks like `file_path` is called from [`Validations::Banner`](https://github.com/zendesk/zendesk_apps_support/blob/olaf/multiple_apps/lib/zendesk_apps_support/validations/banner.rb#L11), so it can't be private.

File: lib/zendesk_apps_support/package.rb (hunk 19 above)
Comment (ocke) [inline lib/zendesk_apps_support/package.rb]: it's only being used by ZAT

File: lib/zendesk_apps_support/package.rb (hunk 26 above)
Comment (ocke) [inline lib/zendesk_apps_support/package.rb]: seems like I missed that one...

File: config/locales/en.yml (hunk 14 above)
Comment (ocke) [inline config/locales/en.yml]: done

**Comment by tobico** [comment]
:+1:

File: lib/zendesk_apps_support/package.rb [hunk 27]
```diff
@@ -6,29 +6,33 @@ module ZendeskAppsSupport
...
     attr_reader :lib_root, :root, :warnings
-    attr_accessor :requirements_only
 
-    def initialize(dir)
-      @root = Pathname.new(File.expand_path(dir))
+    def initialize(dir, cache = true)
+      @root     = Pathname.new(File.expand_path(dir))
       @lib_root = Pathname.new(File.join(@root, 'lib'))
+
+      @cache    = cache #disabled by ZAT for development
```
Comment (danielbreves) [inline lib/zendesk_apps_support/package.rb]: Can you rename this to a boolean sounding name? `is_cached`?

File: lib/zendesk_apps_support/package.rb [hunk 28]
```diff
@@ -42,81 +46,72 @@ def validate
           errors << Validations::Banner.call(self)
         end
 
-        errors.flatten!
+        errors.flatten!.compact!
```
Comment (danielbreves) [inline lib/zendesk_apps_support/package.rb]: When will errors contain `nil`s?

File: lib/zendesk_apps_support/package.rb [hunk 29]
```diff
@@ -131,42 +126,64 @@ def readified_js(app_name, app_id, asset_url_prefix, settings = {}, locale = 'en
...
-      file_exists?('app.js')
+    def requirements_json
+      return nil unless has_requirements?
+      @requirements ||= read_json('requirements.json')
     end
 
-    def has_lib_js?
-      lib_files.any?
+    def no_template
+      if manifest_json['noTemplate'].is_a?(Array)
```
Comment (danielbreves) [inline lib/zendesk_apps_support/package.rb]: Can't we just call `!!no_template_locations` here?

Comment (danielbreves) [inline lib/zendesk_apps_support/package.rb]: Can't we just call `!!no_template_locations` here?

File: spec/package_spec.rb [hunk 30]
```diff
@@ -55,85 +56,331 @@
...
+  let(:package) { ZendeskAppsSupport::Package.new(source) }
+
+  def build_app_source_with_files(files)
+    build_app_source({
+      manifest: manifest,
+      additional_files: files
+    })
+  end
+
+  describe 'test function works as expected' do
```
Comment (danielbreves) [inline spec/package_spec.rb]: Which function?

File: lib/zendesk_apps_support/package.rb (hunk 28 above)
Comment (ocke) [inline lib/zendesk_apps_support/package.rb]: the validation functions return nil if they don't find any errors.

File: lib/zendesk_apps_support/assets/src.js.erb [hunk 31]
```diff
@@ -1,38 +1,32 @@
-(function() {
```
Comment (sandlerr) [inline lib/zendesk_apps_support/assets/src.js.erb]: stupid question, why remove this?

File: lib/zendesk_apps_support/package.rb [hunk 32]
```diff
@@ -42,81 +46,72 @@ def validate
...
         end
 
-        errors.flatten!
+        errors.flatten!.compact!
       end
     end
 
-    def app_js
-      read_file('app.js')
+    def validate!(marketplace: true)
```
Comment (sandlerr) [inline lib/zendesk_apps_support/package.rb]: :-1:
can `validate` and `validate!` have the same default value please?

File: lib/zendesk_apps_support/package.rb [hunk 33]
```diff
@@ -131,42 +126,64 @@ def readified_js(app_name, app_id, asset_url_prefix, settings = {}, locale = 'en
...
+      result
     end
 
-    def has_banner?
-      file_exists?('assets/banner.png')
+    def has_location?
+      manifest_json['location']
+    end
+
+    def customer_css
```
Comment (sandlerr) [inline lib/zendesk_apps_support/package.rb]: let's take the opportunity to give this a less weird variable name? `app_css`?

**Comment by sandlerr** [comment]
looking good

File: lib/zendesk_apps_support/assets/src.js.erb (hunk 31 above)
Comment (ocke) [inline lib/zendesk_apps_support/assets/src.js.erb]: because it's in installed.js.erb

Comment (sandlerr) [inline lib/zendesk_apps_support/assets/src.js.erb]: that's not the same, this was wrapping each individual app, not the whole of installed.js

File: lib/zendesk_apps_support/package.rb (hunk 32 above)
Comment (ocke) [inline lib/zendesk_apps_support/package.rb]: done

File: lib/zendesk_apps_support/assets/src.js.erb (hunk 31 above)
Comment (ocke) [inline lib/zendesk_apps_support/assets/src.js.erb]: done

File: lib/zendesk_apps_support/package.rb (hunk 29 above)
Comment (ocke) [inline lib/zendesk_apps_support/package.rb]: yes, can we do this in a different PR?
There is so much more I want to fix (no_template?)

**Comment by danielbreves** [comment]
:+1:
